data_loader/8_fakenews/fakenews_plugin.py

```diff
@@ -0,0 +1,167 @@
+                        # we collected a full batch and hand it over to the data loader
+                        yield items
+                        items = []
+
+                # forward one page in the api
+                since = data_batch["next"]
+
+                # store the since value, so we can resume later
+                self.key_value_store[section] = since
+
+                # is there more data
+                if data_batch["eof"]:
+                    log.info(f" - end of section {section} reached.")
+                    break
+
+        if items:
+            yield items
+
+        return
+
+    def getSchema(self):
+
+        # get a sample response from the api to read out all available fields
+        # for such a simple api we could also hard code the schema
+
+        data_batch = self.getNewsBatch(section="sport", count=1, since=0)
+        return data_batch["news"][0].keys()
+
+    def getJobId(self) -> str:
+        """Generate a stable ID that changes with the main parameters."""
+        m = hashlib.blake2b(digest_size=20)
+        for v in (self.args.endpoint, self.args.section):
+            m.update(repr(v).encode())
+
+        job_id = m.hexdigest()
+        log.info("Job ID: %r", job_id)
+        return job_id
+
+    def getArguments(self):
+        return [
+            {
+                "name": "section",
+                "display_label": "Section",
+                "help": "Which sections to index",
+                "nargs": "*",
+                "type": "str",
+                "required": True,
+                "advanced": False,
+            },
+            {
+                "name": "endpoint",
+                "display_label": "API Endpoint",
+                "default": "https://fakenews.squirro.com",
+                "help": "API Endpoint",
+                "type": "str",
+                "advanced": True,
+            },
+        ]
+
+    def getNewsBatch(self, section, count=100, since=0):
+        """
+        Helper function to get a single batch of news articles from the API
+        Implements caching to avoid duplicate roundtrips.
+        """
+
+        base_url = f"{self.args.endpoint}/news/{section}"
+        request_params = {"count": count, "since": since}
+        url = base_url + "?" + urllib.parse.urlencode(request_params)
+
+        cache_hit = self.key_value_cache.get(url)
+
+        if cache_hit:
+            return json.loads(cache_hit)
+
+        # call the api
+        response = self.requests.get(url, headers={"Content-Type": "application/json"})
+
+        # throw an exception if the http status is not ok
+        response.raise_for_status()
+
+        # parse the json data into a python dict
+        data_batch = response.json()
+
+        # log.info(json.dumps(data_batch, indent=2))
+
+        # we only store into cache if this is not the last page
+        if not data_batch["eof"]:
+            self.key_value_cache[url] = response.text
```

Comment:
Should we still encourage this? Rerunning will now happen on pipeline level. I'd rather remove this complexity from the data loader plugins.

Reply:
I'll remove it in the simple example. I'll add a comment about the rerunning. But caching might still be useful for more complex API's